test/integration/deployment/deployment_test.go

```diff
@@ -69,3 +70,59 @@ func TestNewDeployment(t *testing.T) {
 		t.Errorf("expected new ReplicaSet last-applied annotation not copied from Deployment %s", deploy.Name)
 	}
 }
+
+func TestDeploymentSelectorImmutability(t *testing.T) {
+	s, closeFn, c := dcSimpleSetup(t)
+	defer closeFn()
+	name := "test-deployment-selector-immutability"
+	ns := framework.CreateTestingNamespace(name, s, t)
+	defer framework.DeleteTestingNamespace(ns, s, t)
+
+	replicas := int32(20)
+	appsV1beta2Tester := &deploymentAppsV1beta2Tester{t: t, c: c, deployment: newAppsV1beta2Deployment(name, ns.Name, replicas)}
+
+	deploy, err := c.AppsV1beta2().Deployments(ns.Name).Create(appsV1beta2Tester.deployment)
+	if err != nil {
+		t.Fatalf("failed to create deployment %s: %v", deploy.Name, err)
+	}
+
+	// Change selector - selectors are IMMUTABLE for all API versions except apps/v1beta1 and extensions/v1beta1
+	oldSelectorLabels := appsV1beta2Tester.deployment.Spec.Selector.MatchLabels
+	newSelectorLabels := map[string]string{"changed_name": "changed_test"}
+	appsV1beta2Tester.deployment.Spec.Selector.MatchLabels = newSelectorLabels
+	deploy, err = c.AppsV1beta2().Deployments(ns.Name).Update(appsV1beta2Tester.deployment)
+	if err != nil {
+		t.Fatalf("failed to update deployment %s: %v", deploy.Name, err)
+	}
+	if !reflect.DeepEqual(deploy.Spec.Selector.MatchLabels, oldSelectorLabels) {
+		t.Errorf("selector should NOT be changed for apps/v1beta2, expected: %v, got: %v", newSelectorLabels, deploy.Spec.Selector.MatchLabels)
+	}
+}
+
+func TestDeploymentSelectorMutability(t *testing.T) {
```

Comment:
I think it's better to combine these functions because they're related. In integration tests there's a non-trivial time cost to set up the apiserver every time.